### PR TITLE
Improved Json Serialization

### LINKED ISSUE #3601: Move away from Newtonsoft.Json to more performant implementation
https://medium.com/justeattakeaway-tech/json-serialization-libraries-performance-tests-b54cbb3cccbb

Probably best candidates:
1. SpanJson
2. Utf8Json
3. System.Text.Json (.net one)

First evaluate:
1. Features
2. Some simple benchmarks of our use cases

Then move:
1. JsonSerializer
2. JSON RPC
3. All the custom converters

### LINKED ISSUE #4314: Migrate to System.Text.Json
Consider migrating from `Newtonsoft.Json` to `System.Text.Json`.

See [devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7](https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7/)

### PR DESCRIPTION
**Pull request by benaadams**
Resolves #4314
Resolves #3601

## Changes

- _List the changes_

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Config/JsonConfigSource.cs [hunk 1]
```diff
@@ -68,20 +68,37 @@ private void LoadJsonConfig(string configFilePath)
...
+                        itemsDict[key] = "true";
+                    }
+                    else if (value.ValueKind == JsonValueKind.False)
+                    {
+                        itemsDict[key] = "false";
+                    }
+                    else
+                    {
+                        itemsDict[key] = configItem.Value.ToString();
+                    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Config/JsonConfigSource.cs:97]: Why not just ToString?

File: src/Nethermind/Nethermind.Core.Test/Builders/TestItem.cs [hunk 2]
```diff
@@ -108,13 +108,8 @@ public static Keccak KeccakFromNumber(int i)
...
         {
-            using var stream = new MemoryStream();
-
-            JsonSerializer.Serialize(stream, value);
-
-            stream.Position = 0;
-
-            return JsonSerializer.Deserialize<T>(stream)!;
+            string data = Newtonsoft.Json.JsonConvert.SerializeObject(value);
+            return Newtonsoft.Json.JsonConvert.DeserializeObject<T>(data)!;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core.Test/Builders/TestItem.cs:112]: Doesn't it use 2x memory now compared to utf8?

File: src/Nethermind/Nethermind.Core/ByteArrayConverter.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            {
+                length = checked((int)reader.ValueSequence.Length);
+                if (length == 0)
+                {
+                    return null;
+                }
+
+                bytes = ArrayPool<byte>.Shared.Rent(length);
+                reader.ValueSequence.CopyTo(bytes);
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/ByteArrayConverter.cs:49]: I don't get that part, why those checks and renting?

File: src/Nethermind/Nethermind.Core/ByteArrayConverter.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            JsonSerializerOptions options)
+        {
+            Convert(writer, bytes, skipLeadingZeros: false);
+        }
+
+        [SkipLocalsInit]
+        public static void Convert(Utf8JsonWriter writer, ReadOnlySpan<byte> bytes, bool skipLeadingZeros = true)
+        {
+            const int maxStackLength = 128;
+            const int stackLength = 256;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/ByteArrayConverter.cs:85]: Why 2x?

File: src/Nethermind/Nethermind.Core/Crypto/PublicKey.cs [hunk 5]
```diff
@@ -136,3 +140,39 @@ public string ToShortString()
...
+            if (bytes is null)
+            {
+                return null;
+            }
+            if (bytes.Length < 64)
+            {
+                var newArray = new byte[64];
+                bytes.AsSpan().CopyTo(newArray.AsSpan(64 - bytes.Length));
+                bytes = newArray;
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Crypto/PublicKey.cs:165]: Other types don't have that safety measure? 
Maybe this should be pre-allocated and passed to `ByteArrayConverter.Convert` ?

File: src/Nethermind/Nethermind.Core/LowerCaseJsonStringEnumConverter.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    }
+}
+
+public class LowerCaseJsonNamingPolicy : JsonNamingPolicy
+{
+    public static LowerCaseJsonNamingPolicy Default { get; } = new();
+
+    public override string ConvertName(string name)
+    {
+        return name.ToLowerInvariant();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/LowerCaseJsonStringEnumConverter.cs:22]: Is there a way to Format name, instead of allocating extra string?

File: src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethTxTraceEntry.cs [hunk 7]
```diff
@@ -15,25 +19,30 @@ public GethTxTraceEntry()
...
         }
 
+        [JsonConverter(typeof(LongRawJsonConverter))]
         public long Pc { get; set; }
 
-        [JsonProperty(PropertyName = "op")]
+        [JsonPropertyName("op")]
         public string? Operation { get; set; }
 
+        [JsonConverter(typeof(LongRawJsonConverter))]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethTxTraceEntry.cs:28]: Downside of this approach is that we need to mark each property everywhere, instead of just relying on proper type conversion

File: src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs [hunk 8]
```diff
@@ -22,8 +25,9 @@ public AccessListItemForRpc(Address address, IReadOnlyCollection<UInt256>? stora
 
         public Address Address { get; set; }
 
-        [JsonProperty(ItemConverterType = typeof(StorageCellIndexConverter))]
-        public UInt256[]? StorageKeys { get; set; }
+        private UInt256[]? _storageKeys;
+        [JsonConverter(typeof(StorageCellIndexConverter))]
+        public UInt256[]? StorageKeys { get => _storageKeys ?? Array.Empty<UInt256>(); set => _storageKeys = value; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs:30]: Maybe not-nullable?

File: src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs [hunk 9]
```diff
@@ -8,12 +8,15 @@
...
-using Newtonsoft.Json;
+using System.Text.Json;
+using System.Text.Json.Serialization;
 
 namespace Nethermind.JsonRpc.Data
 {
     public class AccessListItemForRpc
     {
+        [JsonConstructor]
+        public AccessListItemForRpc() { }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs:19]: can it be private?

File: src/Nethermind/Nethermind.JsonRpc/Data/BlockParameterConverter.cs [hunk 10]
```diff
@@ -1,150 +0,0 @@
...
-
-using System;
-using Nethermind.Blockchain;
-using Nethermind.Blockchain.Find;
-using Nethermind.Core.Crypto;
-using Nethermind.JsonRpc.Modules.Trace;
-using Nethermind.Serialization.Json;
-using Newtonsoft.Json;
-
-namespace Nethermind.JsonRpc.Data
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Data/BlockParameterConverter.cs:12]: Benefit of this approach was that Json serialization/deserialization was decoupled from domain types.

**Review by LukaszRozmej — COMMENTED** [review]
Still found references to remove:


Targets
Occurrences of 'Newtonsoft.json' in Solution
Found usages
<Plugins>
<AccountAbstraction>
<Nethermind.AccountAbstraction.Test>
<Plugins>\<AccountAbstraction>\<Nethermind.AccountAbstraction.Test>\AccountAbstractionRpcModuleTests.cs
using Newtonsoft.Json.Linq;
<Merge>
<Nethermind.JsonRpc.Test>
Nethermind.JsonRpc.Test.csproj
    <PackageReference Include="Newtonsoft.Json" />
<Nethermind.Merge.Plugin.Test>
<Plugins>\<Merge>\<Nethermind.Merge.Plugin.Test>\ExternalRpcIntegrationTests.cs
using Newtonsoft.Json.Linq;
<Mev>
<Nethermind.Mev.Test>
Nethermind.Mev.Test.csproj
    <PackageReference Include="Newtonsoft.Json" />
<Tests>
<Nethermind.Abi.Test>
Json
AbiDefinitionParserTests.cs
using Newtonsoft.Json.Linq;
Nethermind.Abi.Test.csproj
    <PackageReference Include="Newtonsoft.Json" />
<Nethermind.Config.Test>
Nethermind.Config.Test.csproj
    <PackageReference Include="Newtonsoft.Json" />
<Nethermind.Consensus.Test>
Nethermind.Consensus.Test.csproj
    <PackageReference Include="Newtonsoft.Json" />
<Nethermind.Core.Test>
Builders
TestItem.cs
            string data = Newtonsoft.Json.JsonConvert.SerializeObject(value);
            return Newtonsoft.Json.JsonConvert.DeserializeObject<T>(data)!;
<Nethermind.JsonRpc.Test>
Modules
Eth
EthRpcModuleTests.EstimateGas.cs
using Newtonsoft.Json.Linq;
<Nethermind.Overseer.Test>
Framework
AuRaContext.cs
using Newtonsoft.Json.Linq;

**Review by MarekM25 — COMMENTED** [review]
@benaadams do we have any benchmarks? Btw we have one tool for benchmarking JSON-RPC, we plan to check receipts eth_getLogs after optimization, but IG this PR should be checked as well

File: src/Nethermind/Nethermind.Abi/AbiType.TypeMapping.cs [hunk 11]
```diff
@@ -44,19 +48,45 @@ protected static AbiType GetForCSharpType(Type type)
...
         {
-            _typeMappings[typeof(T)] = abiType;
+            lock (_registerLock)
+            {
+                Dictionary<Type, AbiType> typeMappings = new(_typeMappings)
+                {
+                    [typeof(T)] = abiType
+                };
+
+                _typeMappings = typeMappings;
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Nethermind/Nethermind.Abi/AbiType.TypeMapping.cs:65]: A concurrent reader might be reading from the `_typeMappings` instance.

[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Abi/AbiType.TypeMapping.cs:65]: Adds are very infrequent (basically non-existent) so rather than having to pay for a heavier data structure like ConcurrentDictionary or a lock on every read we can use an unprotected raw Dictionary if we replace the pointer entirely on add

File: src/Nethermind/Nethermind.Core.Test/Json/NullableLongConverterTests.cs [hunk 12]
```diff
@@ -3,102 +3,75 @@
...
         public void Test_roundtrip(NumberConversion numberConversion)
         {
-            NullableLongConverter converter = new(numberConversion);
             TestConverter(int.MaxValue, (a, b) => a.Equals(b), converter);
             TestConverter(1L, (a, b) => a.Equals(b), converter);
             TestConverter(0L, (a, b) => a.Equals(b), converter);
         }
 
-        [TestCase((NumberConversion)99)]
-        public void Unknown_not_supported(NumberConversion notSupportedConversion)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core.Test/Json/NullableLongConverterTests.cs:26]: Why removed?

File: src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethLikeTxTrace.cs [hunk 13]
```diff
@@ -21,7 +24,43 @@ public GethLikeTxTrace()
...
+        public override void Write(
+            Utf8JsonWriter writer,
+            GethLikeTxTrace value,
+            JsonSerializerOptions options)
+        {
+            writer.WriteStartObject();
+
+            if (value.StoragesByDepth.Count > 0)
+            {
+                writer.WritePropertyName("structLogs"u8);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethLikeTxTrace.cs:47]: `structLogs` twice?

File: src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethTxTraceEntry.cs [hunk 14]
```diff
@@ -48,4 +57,19 @@ internal void UpdateMemorySize(ulong size)
...
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options) => throw new NotImplementedException();
+
+        public override void Write(
+            Utf8JsonWriter writer,
+            long value,
+            JsonSerializerOptions options)
+        {
+            writer.WriteNumberValue(value);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethTxTraceEntry.cs:72]: Hm... why is this needed? Can't it infer it automatically?

File: src/Nethermind/Nethermind.Evm/TransactionSubstate.cs [hunk 15]
```diff
@@ -47,24 +47,32 @@ public TransactionSubstate(EvmExceptionType exceptionType, bool isTracerConnecte
...
-                        try
-                        {
-                            BigInteger start = Output.Span.Slice(4, 32).ToUnsignedBigInteger();
-                            BigInteger length = Output.Slice((int)start + 4, 32).Span.ToUnsignedBigInteger();
-                            Error = string.Concat("Reverted ",
-                                Output.Slice((int)start + 32 + 4, (int)length).ToArray().ToHexString(true));
-                        }
-                        catch (Exception)
+                        var span = Output.Span;
+                        if (span.Length > 32 + 4)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Evm/TransactionSubstate.cs:51]: This was not there before. Why? And what does this have to do with JSON? Can't you put it on a different PR?

File: src/Nethermind/Nethermind.JsonRpc.Test/Data/SerializationTestBase.cs [hunk 16]
```diff
@@ -16,10 +18,10 @@ public class SerializationTestBase
         protected void TestRoundtrip<T>(T item, Func<T, T, bool>? equalityComparer, JsonConverter<T>? converter = null, string? description = null)
         {
             IJsonSerializer serializer = BuildSerializer();
-            if (converter is not null)
-            {
-                serializer.RegisterConverter(converter);
-            }
+            //if (converter is not null)
+            //{
```
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.Test/Data/SerializationTestBase.cs:22]: Missed? Why not remove?

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs [hunk 17]
```diff
@@ -529,7 +529,7 @@ public async Task Trace_transaction_with_error_reverted()
...
...28db452defc0214bc36b22\",\"transactionPosition\":0,\"type\":\"create\",\"error\":\"Reverted\"},{\"action\":{\"callType\":\"call\",\"from\":\"0xd6a48bcd4c5ad5adacfab677519c25ce7b2805a5\",\"gas\":\"0x2dcd\",\"input\":\"0x\",\"to\":\"0x6b5887043de753ecfa6269f947129068263ffbe2\",\"value\":\"0x0\"},\"blockHash\":\"0xfa74e932520ee416ecb12171c115b3ad14112ffd2d612646ceaff69e54e06a94\",\"blockNumber\":18,\"result\":{\"gasUsed\":\"0x0\",\"output\":\"0x\"},\"subtraces\":0,\"traceAddress\":[0],\"transactionHash\":\"0x787616b8756424622f162fc3817331517ef941366f28db452defc0214bc36b22\",\"transactionPosition\":0,\"type\":\"call\"},{\"action\":{\"callType\":\"call\",\"from\":\"0xd6a48bcd4c5ad5adacfab677519c25ce7b2805a5\",\"gas\":\"0x2d56\",\"input\":\"0x\",\"to\":\"0x6b5887043de753ecfa6269f947129068263ffbe2\",\"value\":\"0x0\"},\"blockHash\":\"0xfa74e932520ee416ecb12171c115b3ad14112ffd2d612646ceaff69e54e06a94\",\"blockNumber\":18,\"result\":{\"gasUsed\":\"0x0\",\"output\":\"0x\"},\"subtraces\":0,\"traceAddress\":[1],\"transactionHash\":\"0x787616b8756424622f162fc3817331517ef941366f28db452defc0214bc36b22\",\"transactionPosition\":0,\"type\":\"call\"}]", serialized, serialized.Replace("\"", "\\\""));
```
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs:532]: Why is it different?

File: src/Nethermind/Nethermind.JsonRpc/IJsonRpcProcessor.cs [hunk 18]
```diff
@@ -1,16 +1,14 @@
...
-using Nethermind.JsonRpc.Modules;
-using Nethermind.Serialization.Json;
+using System.IO.Pipelines;
 
 namespace Nethermind.JsonRpc
 {
     public interface IJsonRpcProcessor
     {
-        IAsyncEnumerable<JsonRpcResult> ProcessAsync(TextReader request, JsonRpcContext context);
+        IAsyncEnumerable<JsonRpcResult> ProcessAsync(PipeReader stream, JsonRpcContext context);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc/IJsonRpcProcessor.cs:12]: What is the advantage of PipeReader?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/BlockForRpc.cs [hunk 19]
```diff
@@ -83,22 +86,22 @@ public BlockForRpc(Block block, bool includeFullTransactionData, ISpecProvider s
...
-    [JsonProperty(NullValueHandling = NullValueHandling.Include)]
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
     public Bloom LogsBloom { get; set; }
     public Address Miner { get; set; }
     public Keccak MixHash { get; set; }
 
     public bool ShouldSerializeMixHash() => !_isAuRaBlock && MixHash is not null;
 
-    [JsonProperty(NullValueHandling = NullValueHandling.Include)]
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
```
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/BlockForRpc.cs:99]: Not `Never`?

File: src/Nethermind/Nethermind.Evm/TransactionSubstate.cs (hunk 15 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/TransactionSubstate.cs:51]: Was annoying me when debugging. Have added separate PR https://github.com/NethermindEth/nethermind/pull/5727

File: src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs (hunk 9 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.JsonRpc/Data/AccessListItemForRpc.cs:19]: In .NET 7 and earlier versions, the [JsonConstructor] attribute can only be used with public constructors.

**Comment by benaadams** [comment]
Replaced by https://github.com/NethermindEth/nethermind/pull/6152
